### DISCUSSION TITLE
How to translate cursor position into spot in JSON

### DISCUSSION THREAD

**Discussion by ZakMiller** [ueberdosis/tiptap#806] [General Questions]
I'm programmatically insert text at the cursor position.
I achieved this by getting the cursor position:
```
 editor: new Editor({
        onTransaction: ({ state }) => {
          this.cursorPosition = state.selection.anchor - 1;
        },
})
```
And then looping through the JSON paragraphs and content to find the appropriate spot (adding up the text length). This is working, but seems prone to error. Is there a better way to do this?

Related: after inserting text at the cursor programmatically, the cursor is removed from the editor. I'd like to add it back. I'm doing that using a similar mechanism:
```
this.editor.setSelection = priorCursorPosition + insertedText.length
```
Again, is there a better way of achieving this?

**Comment by hanspagel** [comment]
Oh, there is an `insertText` API method. I think that’s what you’re looking for. Try the following:

1. Create a new extension (duplicate one that’s more or less what you want to do, e.g. `Bold.js`)
2. Import insertText form `tiptap-commands`
3. Add the call to insertText in the `commands()` section.
4. Register your new extension with tiptap `new Editor({extensions: [new CustomExtension]})`

Hope that helps!

**Reply by ZakMiller** [reply]
Hey, thank you for the response!

I don't just want to insert text, though. I should have mentioned. I want to insert text with a particular mark (so the text is clearly differentiated). I have this working (including with inserting multiple paragraphs at once) via editing the JSON and then setting it. Really the only issue is figuring out how the cursor position I can get from the editor matches up with the JSON. I have it working most of the time but it seems inconsistent (in particular when it comes to newlines).

**Reply by hanspagel** [reply]
Then you should use the ProseMirror API to insert Nodes, see more here:
https://discuss.prosemirror.net/t/example-of-inserting-nodes/234/5

I don’t think there is another way to get this consistent.
